examples/src/main/java/io/grpc/examples/experimental/CompressingHelloWorldServer.java

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright 2015 The gRPC Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.examples.experimental;
+
+import java.io.IOException;
+import java.util.logging.Logger;
+
+import io.grpc.CompressorRegistry;
+import io.grpc.DecompressorRegistry;
+import io.grpc.Server;
+import io.grpc.ServerBuilder;
+import io.grpc.examples.helloworld.GreeterGrpc;
+import io.grpc.examples.helloworld.HelloReply;
+import io.grpc.examples.helloworld.HelloRequest;
+import io.grpc.stub.StreamObserver;
+
+/**
+ * Server that manages startup/shutdown of a {@code Greeter} server.
+ */
+public class CompressingHelloWorldServer {
+  private static final Logger logger = Logger.getLogger(CompressingHelloWorldServer.class.getName());
+
+  private Server server;
+
+  private void start() throws IOException {
+    /* The port on which the server should run */
+    int port = 50051;
+    server = ServerBuilder.forPort(port)
+        .compressorRegistry(CompressorRegistry.getDefaultInstance())
+        .decompressorRegistry(DecompressorRegistry.getDefaultInstance())
+        .intercept(new CompressionServerInterceptor())
```

Comment:
It needs to be made clear why you are using an interceptor. Frequently individual methods that know they want compression could enable it by casting StreamObserver to ServerCallStreamObserver and calling `setCompression("gzip")`.

I assume you're using an interceptor here to enable it for all methods.

Reply:
Thanks a lot for the alternative. I could not find a recommended way to enable server response compression and figured out to use ServerInterceptor to do that.

Yes, I want to enable it in one place for all the methods.

I tried ServerCallStreamObserver.setCompression("gzip") and it works. I just have some questions:

1. Is it always safe to cast StreamObserver to ServerCallStreamObserver?
2. Can we do it in one place?
3. Do we need to do explicit check to see if the client accept/prefer compression? If yes, how?
That's related to one of my previous replies.

Thanks again.

Reply:
> Is it always safe to cast StreamObserver to ServerCallStreamObserver?

Yes, it is safe for the StreamObserver passed to the server by gRPC. It is really unfortunate that it is hard to discover and awkward and ugly, but it's what we were forced to do to 1) add features and 2) not break existing code.

> Can we do it in one place?

To do it in one place, use the interceptor. If you want to configure it per-method, use the ServerCallStreamObserver.

> Do we need to do explicit check to see if the client accept/prefer compression? If yes, how?

No. [We do that in gRPC.](https://github.com/grpc/grpc-java/blob/v1.18.0/core/src/main/java/io/grpc/internal/ServerCallImpl.java#L97) Although it does seem the Javadoc should be improved to say it is safe.

Reply:
Sent out #5401 to improve documentation.

Reply:
Thanks a lot for the explanations and documentation improvement. Provide two examples to show both cases: PerMethod and AllMethods.